test/integration/shadow_policy_integration_test.cc

```diff
@@ -85,15 +89,323 @@ class ShadowPolicyIntegrationTest : public testing::TestWithParam<Network::Addre
+  auto& encoder = result.first;
+  auto response = std::move(result.second);
+
+  FakeHttpConnectionPtr fake_upstream_connection_0;
+  FakeStreamPtr upstream_request_0;
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_0));
+  ASSERT_TRUE(fake_upstream_connection_0->waitForNewStream(*dispatcher_, upstream_request_0));
+  FakeHttpConnectionPtr fake_upstream_connection_1;
+  FakeStreamPtr upstream_request_1;
+  ASSERT_TRUE(fake_upstreams_[1]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_1));
+  ASSERT_TRUE(fake_upstream_connection_1->waitForNewStream(*dispatcher_, upstream_request_1));
+  ASSERT_TRUE(upstream_request_0->waitForHeadersComplete());
+  ASSERT_TRUE(upstream_request_1->waitForHeadersComplete());
+  EXPECT_EQ(upstream_request_0->headers().get(Http::LowerCaseString("potato"))[0]->value(),
+            "salad");
+  EXPECT_EQ(upstream_request_1->headers().get(Http::LowerCaseString("potato"))[0]->value(),
+            "salad");
+
+  codec_client_->sendReset(encoder);
+
+  ASSERT_TRUE(upstream_request_0->waitForReset());
+  ASSERT_TRUE(upstream_request_1->waitForReset());
+  ASSERT_TRUE(fake_upstream_connection_0->close());
+  ASSERT_TRUE(fake_upstream_connection_1->close());
+  ASSERT_TRUE(fake_upstream_connection_0->waitForDisconnect());
+  ASSERT_TRUE(fake_upstream_connection_1->waitForDisconnect());
+
+  EXPECT_FALSE(upstream_request_0->complete());
+  EXPECT_FALSE(upstream_request_1->complete());
+  EXPECT_FALSE(response->complete());
+
+  cleanupUpstreamAndDownstream();
+
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_cx_total")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_0.upstream_cx_total")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_rq_tx_reset")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_0.upstream_rq_tx_reset")->value(), 1);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_1.upstream_rq_completed")->value(), 0);
+  EXPECT_EQ(test_server_->counter("cluster.cluster_0.upstream_rq_completed")->value(), 0);
+}
+
+TEST_P(ShadowPolicyIntegrationTest, RequestMirrorPolicyWithMainUpstreamReset) {
+  if (!streaming_shadow_) {
+    GTEST_SKIP() << "Not applicable for non-streaming shadows.";
+  }
+  autonomous_upstream_ = false;
+  /*
+  TestScopedRuntime scoped_runtime;
+  scoped_runtime.mergeValues({{"envoy.reloadable_features.remember_shadow", "true"}});*/
+  initialConfigSetup("cluster_1", "");
+  initialize();
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  Http::TestRequestHeaderMapImpl request_headers = default_request_headers_;
+  request_headers.addCopy("potato", "salad");
+  std::pair<Http::RequestEncoder&, IntegrationStreamDecoderPtr> result =
+      codec_client_->startRequest(request_headers, false);
+  auto response = std::move(result.second);
+
+  FakeHttpConnectionPtr fake_upstream_connection_0;
+  FakeStreamPtr upstream_request_0;
```

Comment:
nit: can these be named to say which is the shadow_request, main_request, etc.?

Reply:
Done.